huaweicloud/services/dew/resource_huaweicloud_kps_keypair.go

```diff
@@ -0,0 +1,365 @@
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := config.NewKmsClient(c, region)
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating KMS v3 client: %s", err)
+	}
+
+	updateErr := updateDesc(d, client)
+	if updateErr != nil {
+		return updateErr
+	}
+
+	return resourceKeypairRead(ctx, d, meta)
+}
+
+func resourceKeypairDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := config.NewKmsClient(c, region)
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating KMS v3 client: %s", err)
+	}
+
+	deleteOpts := &model.DeleteKeypairRequest{
+		KeypairName: d.Id(),
+	}
+
+	_, err = client.DeleteKeypair(deleteOpts)
+	if err != nil {
+		return fmtp.DiagErrorf("Error deleting keypair: %s", err)
+	}
+
+	return nil
+}
+
+func buildCreateParams(d *schema.ResourceData) (*model.CreateKeypairRequest, error) {
+	var importPublicKey *string
+	if v, ok := d.GetOk("public_key"); ok {
+		importPublicKey = utils.String(v.(string))
+	}
+
+	createOpts := &model.CreateKeypairRequest{
+		Body: &model.CreateKeypairRequestBody{
+			Keypair: &model.CreateKeypairAction{
+				Name:      d.Get("name").(string),
+				PublicKey: importPublicKey,
+			},
+		},
+	}
+
+	if v, ok := d.GetOk("type"); ok {
+		var actionType model.CreateKeypairActionType
+		err := actionType.UnmarshalJSON([]byte(v.(string)))
+		if err != nil {
+			return nil, fmtp.Errorf("Error parsing the argument %q: %s", "type", err)
+		}
+		createOpts.Body.Keypair.Type = &actionType
+	}
+
+	if v, ok := d.GetOk("scope"); ok {
+		var actionScope model.CreateKeypairActionScope
+		err := actionScope.UnmarshalJSON([]byte(v.(string)))
+		if err != nil {
+			return nil, fmtp.Errorf("Error parsing the argument %q: %s", "scope", err)
+		}
+		createOpts.Body.Keypair.Scope = &actionScope
+	}
+
+	if v, ok := d.GetOk("encryption_type"); ok {
+		t := v.(string)
+		var encryptionType model.EncryptionType
+		err := encryptionType.UnmarshalJSON([]byte(t))
+		if err != nil {
+			return nil, fmtp.Errorf("Error parsing the argument %q: %s", "encryption_type", err)
+		}
+
+		keyProtection := model.KeyProtection{
+			Encryption: &model.Encryption{
+				Type: &encryptionType,
+			},
+		}
+
+		//the kms key name is required when encryption_type="kms"
+		if k, kmsExist := d.GetOk("encryption_kms_key_name"); t == "kms" && !kmsExist {
+			return nil, fmtp.Errorf("encryption_kms_key_name is mandatory when the encryption_type is kms")
+		} else {
+			keyProtection.Encryption.KmsKeyName = utils.String(k.(string))
+		}
+
+		createOpts.Body.Keypair.KeyProtection = &keyProtection
+	}
+
+	return createOpts, nil
+}
+
+func parseEncodeValue(b []byte, err error) (*string, error) {
+	if err != nil {
+		return nil, err
+	}
+
+	var rst *string
+	err = json.NewDecoder(bytes.NewReader(b)).Decode(&rst)
+	if err != nil {
+		return nil, err
+	}
+
+	return rst, nil
+}
+
+func getKeyFilePath(d *schema.ResourceData) string {
```

Comment:
it's unnecessary to difine this func, use d.Get().(string) is just fine

Reply:
done